feat(settings): Add service hook settings

dcramer opened the pull request:
- Make 'events' optional for service hook endpoints
- Add service hook stats endpoint
- Add service hook list
- Add service hook creation
- Add service hook details (and edit)

dcramer commented:
This exposes basically nothing (just the URL).

I want to:

- add events (multi checkbox)
- add read-only signing details 
- add graph (similar to client key details page, which hasn't been converted yet i think)

getsentry-danger commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

dcramer commented on src/sentry/static/sentry/app/data/forms/serviceHookSettings.jsx:
```diff
@@ -0,0 +1,24 @@
+import {t} from '../../locale';
+
+// Export route to make these forms searchable by label/help
+export const route = '/settings/organization/:orgId/project/:projectId/hooks/:hookId/';
```
we should revisit these routes as they're quite verbose

dcramer commented on src/sentry/static/sentry/app/data/forms/serviceHookSettings.jsx:
```diff
@@ -0,0 +1,24 @@
+    fields: [
+      {
+        name: 'url',
+        type: 'string',
```
a url type (or validators) would be nice and not an uncommon need

dcramer commented on src/sentry/static/sentry/app/routes.jsx:
```diff
@@ -239,6 +239,22 @@ const projectSettingsRoutes = [
       import(/*webpackChunkName: "ProjectTeams"*/ './views/settings/project/projectTeams')}
     component={errorHandler(LazyLoad)}
   />,
+  <Route
```
this seems like an awfully verbose way of declaring routes

billyvg replied:
Yeah, we can now get rid of `key` by using `React.Fragment`, and create a `LazyRoute`, but we use `name` for breadcrumbs, and dynamic imports require static strings

dcramer commented on src/sentry/static/sentry/app/views/settings/project/projectServiceHookDetails.jsx:
```diff
@@ -0,0 +1,105 @@
+import serviceHookSettingsFields from '../../../data/forms/serviceHookSettings';
+import SettingsPageHeader from '../components/settingsPageHeader';
+
+const TOAST_DURATION = 10000;
```
this was copy pasted from org settings

billyvg replied:
Toasts now have a default duration configured in a constants file

dcramer commented on src/sentry/static/sentry/app/views/settings/project/projectServiceHookDetails.jsx:
```diff
@@ -0,0 +1,105 @@
+        saveOnBlur
+        allowUndo
+        initialData={initialData}
+        onSubmitSuccess={(change, model, id) => {
```
this was copy pasted from org settings

billyvg replied:
There's a PR that refactors and provides this by default

dcramer commented on src/sentry/static/sentry/app/views/settings/project/projectServiceHookDetails.jsx:
```diff
@@ -0,0 +1,105 @@
+  },
+});
+
+export default class ProjectServiceHookDetails extends AsyncView {
```
can we just make an AsyncSettingsView or something similar? (it could have a getTitle() which does the Service Hook Details part here)

billyvg replied:
Yeah

dcramer commented:
i added an initial skeleton for creating a new hook and removed some code that I think wasnt needed

a few things I have to sort out:

1. the events field type (its a multi checkbox)
2. save + error doesn't enable the submit button by default (this might be a core bug in the new Form stuff)
3. for all i know i broke the edit page that used to exist
4. before we merge any of this, we'll want to make sure its all behind feature switches

dcramer commented:
![screen shot 2018-04-17 at 10 45 38 am](https://user-images.githubusercontent.com/23610/38887024-7f68dc2e-422c-11e8-8410-0a1d61756f7e.png)

![screen shot 2018-04-17 at 10 45 43 am](https://user-images.githubusercontent.com/23610/38887030-82512a18-422c-11e8-80f3-7a8a450fd2ab.png)

dcramer commented:
/cc @mnoble

dcramer commented:
I'm pretty content with what we expose at this point. It needs some UI cleanup as theres some issues w/ the newer components.

mnoble commented on tests/acceptance/test_project_servicehooks.py:
```diff
@@ -0,0 +1,54 @@
+
+        self.login_as(self.user)
+        self.path1 = '/settings/{}/{}/hooks/'.format(self.org.slug, self.project.slug)
+        self.path2 = '/settings/{}/{}/hooks/new/'.format(self.org.slug, self.project.slug)
```
It'd be easier to read what's going on below if these were `self.hooks_path` and `self.new_hook_path`

mnoble commented on src/sentry/api/endpoints/project_servicehook_details.py:
```diff
@@ -67,12 +68,16 @@ def put(self, request, project, hook_id):
         result = validator.object
```
How does this endpoint work with OAuth? Is `authentication_classes` and `permission_classes` not being defined mean this won't be accessible via oauth apps?

dcramer replied:
@mnoble by default we support "all forms of authentication"

```python
    authentication_classes = DEFAULT_AUTHENTICATION
```

```python
DEFAULT_AUTHENTICATION = (
    TokenAuthentication, ApiKeyAuthentication, SessionAuthentication, )
```

Permission defaults to none, but we override that with the base classes.

```python
    permission_classes = (NoPermission, )
```

becomes

```python
    permission_classes = (ProjectPermission, )
```

mnoble replied:
Got it 👍

billyvg commented on src/sentry/static/sentry/app/views/settings/project/projectServiceHookDetails.jsx:
```diff
@@ -0,0 +1,96 @@
+        <Panel>
+          <PanelHeader>{t('Delete Hook')}</PanelHeader>
+          <PanelBody>
+            <TextBlock>Removing this hook is immediate and permanent.</TextBlock>
```
* `t()`
* You can use <Field label={...} help={...}><Button /></Field>

billyvg replied:
Oops markdown parsed out my code: `<Field label={...} help={...}><Button /></Field>`

billyvg commented on src/sentry/static/sentry/app/views/settings/project/projectServiceHooks.jsx:
```diff
@@ -0,0 +1,161 @@
+    let {orgId, projectId, hook} = this.props;
+    return (
+      <PanelItem>
+        <Box flex="1">
```
I think you probably want a `<Field>` here too?

dcramer replied:
O.o

dcramer replied:
Can you explain what you mean? I dont understand why I'd want a field here. What I want is tabular data in a grid. There just happens to only be two columns.

dcramer replied:
Also this is where I'd prefer to use the new BooleanField, but fitting a form in here is fairly illogical

billyvg replied:
I guess there's no screenshot for this so I can't tell what it looks like, but if you used field it could give a more consistent design

billyvg commented on src/sentry/static/sentry/app/views/settings/project/projectServiceHooks.jsx:
```diff
@@ -0,0 +1,161 @@
+  renderResults() {
+    let {orgId, projectId} = this.props.params;
+
+    return [
```
`<React.Fragment>` instead of array here

dcramer replied:
whats the actual reason for fragment vs array?

billyvg replied:
Functionally they're both the same, but using Fragment is a bit cleaner since the reason we use an array like this previously was as a work-around to avoid having to wrap this in an extra DOM node.

dcramer replied:
If there's no negative to the array though, and React supports that natively (like children), why wouldn't we prefer the array? It's one less component to deal w/

billyvg replied:
imo because it should look more like a component. If for some reason we did need to wrap this with a `<div>`, you wouldn't it as an array.

dcramer commented:
![screen shot 2018-04-18 at 9 53 16 am](https://user-images.githubusercontent.com/23610/38946451-57230ba0-42ee-11e8-8637-cd84d4076c07.png)

remaining concerns before I'd like to merge:

1. lol the button wtf
2. I'd prefer to use the new boolean input instead of a checkbox on the list
3. the padding is all messed up on the text block

billyvg commented on src/sentry/static/sentry/app/routes.jsx:
```diff
@@ -325,6 +323,30 @@ const projectSettingsRoutes = (
+        import(/*webpackChunkName: "ProjectServiceHooks"*/ './views/settings/project/projectServiceHooks')}
+      component={errorHandler(LazyLoad)}
+    />
+    <Route
```
Breadcrumbs follow the route hierarchy, so these should be nested if you wanted the breadcrumbs to reflect that

dcramer replied:
Yeah I figured as much -- I just didnt want to deal w/ creating even more boilerplate components just to nest them.
